tmva/tmva/inc/TMVA/DNN/RMSProp.h

```diff
@@ -0,0 +1,247 @@
+
+   for (size_t i = 0; i < layersNSlices; i++) {
+      size_t weightsNSlices = (layers[i]->GetWeights()).size();
+
+      for (size_t j = 0; j < weightsNSlices; j++) {
+         Matrix_t &currentWeights = layers[i]->GetWeightsAt(j);
+         size_t weightsNRows = currentWeights.GetNrows();
+         size_t weightsNCols = currentWeights.GetNcols();
+
+         fPastSquaredWeightGradients[i].emplace_back(weightsNRows, weightsNCols);
+         fWeightUpdates[i].emplace_back(weightsNRows, weightsNCols);
+         initialize<Architecture_t>(fPastSquaredWeightGradients[i][j], EInitialization::kZero);
+         initialize<Architecture_t>(fWeightUpdates[i][j], EInitialization::kZero);
+      }
+
+      size_t biasesNSlices = (layers[i]->GetBiases()).size();
+
+      for (size_t j = 0; j < biasesNSlices; j++) {
+         Matrix_t &currentBiases = layers[i]->GetBiasesAt(j);
+         size_t biasesNRows = currentBiases.GetNrows();
+         size_t biasesNCols = currentBiases.GetNcols();
+
+         fPastSquaredBiasGradients[i].emplace_back(biasesNRows, biasesNCols);
+         fBiasUpdates[i].emplace_back(biasesNRows, biasesNCols);
+         initialize<Architecture_t>(fPastSquaredBiasGradients[i][j], EInitialization::kZero);
+         initialize<Architecture_t>(fBiasUpdates[i][j], EInitialization::kZero);
+      }
+   }
+}
+
+//______________________________________________________________________________
+template <typename Architecture_t, typename Layer_t, typename DeepNet_t>
+auto TRMSProp<Architecture_t, Layer_t, DeepNet_t>::printMatrix(const typename Architecture_t::Matrix_t &A, const std::string name)
+-> void
+{
+  std::cout << name << "\n";
+  for (size_t i = 0; i < (size_t) A.GetNrows(); ++i) {
+    for (size_t j = 0; j < (size_t) A.GetNcols(); ++j) {
+        std::cout << A(i, j) << " ";
+    }
+    std::cout << "\n";
+  }
+  std::cout << "********\n";
+}
+
+//_________________________________________________________________________________________________
+template <typename Architecture_t, typename Layer_t, typename DeepNet_t>
+auto TRMSProp<Architecture_t, Layer_t, DeepNet_t>::UpdateWeights(size_t layerIndex, std::vector<Matrix_t> &weights,
+                                                             const std::vector<Matrix_t> &weightGradients) -> void
+{
+   std::vector<Matrix_t> &currentLayerPastSquaredWeightGradients = this->GetPastSquaredWeightGradientsAt(layerIndex);
+   std::vector<Matrix_t> &currentLayerWeightUpdates = this->GetWeightUpdatesAt(layerIndex);
+   
+   // std::cout<<"Layer no : "<<layerIndex<<"\n";
+   for (size_t k = 0; k < currentLayerPastSquaredWeightGradients.size(); k++) {
+      
+      // accumulation matrix used for temporary storing of the current accumulation
+      Matrix_t accumulation(currentLayerPastSquaredWeightGradients[k].GetNrows(),
+                            currentLayerPastSquaredWeightGradients[k].GetNcols());
+      
+      // Vt = rho * Vt-1 + (1-rho) * currentSquaredWeightGradients
```

Comment:
nice! i like that you've put the equations there explicitely. otherwise nobody could follow. :+1: